src/NuGet.Clients/NuGet.CommandLine/MsBuildToolset.cs

```diff
@@ -0,0 +1,173 @@
+    public class MsBuildToolset: IComparable
+    {
+        private readonly DateTime _installDate = DateTime.MinValue;
+        private Version _parsedToolsVersion;
+
+        /// <summary>
+        /// This constructor services pre-v15.1 and non-Windows toolsets
+        /// </summary>
+        public MsBuildToolset(string version, string path)
+        {
+            Version = version;
+            Path = path;
+        }
+
+        /// <summary>
+        /// This constructor services v15.1+ toolsets
+        /// </summary>
+        internal MsBuildToolset(ISetupInstance sxsToolset)
+        {
+            Path = GetMsBuildDirFromVsDir(sxsToolset.GetInstallationPath());
+            Version = GetMsBuildVersionFromMsBuildDir(Path);
+            _installDate = ConvertFILETIMEToDateTime(sxsToolset.GetInstallDate());
+        }
+
+        /// <summary>
+        /// This constructor is for testing purposes only
+        /// </summary>
+        public MsBuildToolset(string version, string path, DateTime installDate)
+        {
+            Version = version;
+            Path = path;
+            _installDate = installDate;
+        }
+
+        public Version ParsedVersion
+        {
+            get
+            {
+                if (_parsedToolsVersion == null)
+                {
+                    if (!System.Version.TryParse(Version, out _parsedToolsVersion))
+                    {
+                        _parsedToolsVersion = new Version(0, 0);
+                    }
+                }
+
+                return _parsedToolsVersion;
+            }
+        }
+
+        public string Version { get; private set; }
+
+        public string Path { get; private set; }
+
+        public int CompareTo(object obj)
+        {
+            if (Object.ReferenceEquals(obj, null))
+            {
+                return 1;
+            }
+
+            var rhs = obj as MsBuildToolset;
+            if (obj == null)
+            {
+                throw new ArgumentException("Comparison object not of correct type");
+            }
+
+            // Compare versions
+            var comparison = this.ParsedVersion.Major.CompareTo(rhs.ParsedVersion.Major);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            comparison = this.ParsedVersion.Minor.CompareTo(rhs.ParsedVersion.Minor);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            comparison = this.ParsedVersion.Build.CompareTo(rhs.ParsedVersion.Build);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            comparison = this.ParsedVersion.Revision.CompareTo(rhs.ParsedVersion.Revision);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            // Versions equal; compare by install date/time
+            return this._installDate > rhs._installDate ? 1 : -1;
```

Comment:
Make ``_installDate`` a public property, it looks like the goal of the class is just to hold this info, I don't see why one critical part of that should be private.

Also, this should just use DateTime's comparison. InstallDate.CompareTo(rhs.InstallDate)   This check misses the case where both are equal (should that very rare event ever happen)